File: workflow-api/src/main/scala/com/hashmap/haf/workflow/entity/BaseSqlEntity.scala
package com.hashmap.haf.workflow.entity

import java.util.UUID
import javax.persistence.Column
import javax.persistence.Id
import javax.persistence.MappedSuperclass

import com.hashmap.haf.workflow.util.UUIDConverter

@MappedSuperclass
abstract class BaseSqlEntity[D] extends BaseEntity[D]{

  @Id
  @Column(name = "id")
  protected var id: String = _


  override def getId: UUID = {
    if (id == null) return null
    UUIDConverter.fromString(id)
  }

  override def setId(id: UUID) = {
    this.id = UUIDConverter.fromTimeUUID(id)
  }

  protected def toUUID(src: String): UUID = UUIDConverter.fromString(src)

  protected def toString(timeUUID: UUID): String = UUIDConverter.fromTimeUUID(timeUUID)

}

File: metadata-api/src/main/java/com/hashmap/haf/metadata/config/entity/BaseSqlEntity.java
package com.hashmap.haf.metadata.config.entity;

import com.fasterxml.uuid.Generators;
import com.hashmap.haf.metadata.config.constants.ModelConstants;
import com.hashmap.haf.metadata.config.utils.UUIDConverter;
import lombok.Data;

import javax.persistence.Column;
import javax.persistence.Id;
import javax.persistence.MappedSuperclass;
import java.util.UUID;

@Data
@MappedSuperclass
public abstract class BaseSqlEntity<D> implements BaseEntity<D> {

    @Id
    @Column(name = ModelConstants.ID_PROPERTY)
    protected String id;

    @Override
    public UUID getId() {
        if (id == null) {
            return Generators.timeBasedGenerator().generate();
        }
        return UUIDConverter.fromString(id);
    }

    public void setId(UUID id) {
        this.id = UUIDConverter.fromTimeUUID(id);
    }

    protected UUID toUUID(String src){
        return UUIDConverter.fromString(src);
    }

    protected String toString(UUID timeUUID){
        return UUIDConverter.fromTimeUUID(timeUUID);
    }

}
